cpp/src/arrow/array/util.cc

```diff
@@ -862,6 +854,12 @@ class RepeatedArrayFactory {
 
   template <typename OffsetType>
   Status CreateOffsetsBuffer(OffsetType value_length, std::shared_ptr<Buffer>* out) {
+    if (length_ > std::numeric_limits<OffsetType>::max()) {
```

Comment:
... because what you need to check here is not `length_` but `value_length`, however `value_length` is already cast to OffsetType here